### PR TITLE
fix: align HEX numeric conversion with MySQL

### LINKED ISSUE #28457: [Compatibility]: HEX misclassifies floating-point, negative DECIMAL, Boolean, and prepared arguments
### Is there an existing issue for the same bug?

- [x] I have checked the existing issues.

No issue was found for `HEX()` numeric type dispatch. #28387 concerns bounded result metadata rather than computed values.

### Branch Name

main

### Commit ID

`a9d3b8b857c44f1249e682c91d4d62a914eb7d20`

### Other Environment Information

- Local standalone MatrixOne built from the commit above on macOS arm64
- Reference: MySQL 8.3.0

### Actual Behavior

`HEX()` does not preserve MySQL's numeric semantics for several source types:

```sql
SELECT HEX(CAST(15.5 AS DOUBLE)), HEX(CAST(-15.5 AS DOUBLE));
-- MatrixOne: 10, 0
-- MySQL:     F,  FFFFFFFFFFFFFFF1

SELECT HEX(CAST(-15.5 AS DECIMAL(4,1)));
-- MatrixOne: 0
-- MySQL:     FFFFFFFFFFFFFFF0

SELECT HEX(TRUE), HEX(FALSE);
-- MatrixOne: 31, 30
-- MySQL:     1,  0
```

The DOUBLE behavior occurs for both scalar expressions and DOUBLE/FLOAT columns. Positive fractional DOUBLE values are rounded in MatrixOne but truncated in MySQL's `HEX()` numeric context. Negative DOUBLE/DECIMAL values reach MatrixOne's floating overload and produce `0` instead of the hexadecimal two's-complement integer.

Boolean values take the string path in MatrixOne, so it hex-encodes the characters `"1"`/`"0"` as `31`/`30` rather than formatting the numeric values.

The parameter-marker path is also classified as text:

```sql
PREPARE s FROM 'SELECT HEX(?)';
SET @v=15.5;
EXECUTE s USING @v;
-- MatrixOne: 31352E35  (bytes of "15.5")
-- MySQL:     10
```

All scalar, column, and fresh server-side PREPARE forms were repeated three times. CTAS preserves the same incorrect values, so the difference is not limited to display metadata.

The implementation has direct FLOAT/DOUBLE overloads that call `uint64(math.Round(v))`; this both rounds positive values differently from MySQL and converts negative floating values directly to unsigned. There is no Boolean numeric overload, and an unresolved prepared marker reaches the string overload.

### Expected Behavior

`HEX()` should use MySQL-compatible type-specific numeric conversion:

- FLOAT/DOUBLE inputs should be converted according to MySQL's `HEX()` integer context, including negative values.
- DECIMAL negative values should retain their signed integer representation.
- Boolean expressions should be treated as numeric 0/1.
- Prepared numeric/Boolean values should not be hex-encoded as their text spelling.

### Steps to Reproduce

```sql
CREATE DATABASE codex_hex_numeric;
USE codex_hex_numeric;
CREATE TABLE t(id INT PRIMARY KEY, d DECIMAL(6,2), f DOUBLE, b BOOLEAN);
INSERT INTO t VALUES
  (1,15.5,15.5,TRUE),
  (2,-15.5,-15.5,FALSE);

SELECT id,HEX(d),HEX(f),HEX(b) FROM t ORDER BY id;

PREPARE s FROM 'SELECT HEX(?)';
SET @v=15.5;
EXECUTE s USING @v;
DEALLOCATE PREPARE s;
```

### Controls

- Signed and unsigned integer inputs agree with MySQL.
- String input intentionally uses byte encoding: `HEX('15.5')` agrees in both engines.
- Positive DECIMAL inputs such as `DECIMAL '15.5'` agree for the tested values.
- `BIN()` and `CONV()` were checked separately and do not share the same FLOAT/DOUBLE value error in this matrix.

### PR DESCRIPTION
**Pull request by ck89119**
## What type of PR is this?

- [ ] API-change
- [x] BUG
- [ ] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

Fixes #28457

## What this PR does / why we need it:

Aligns `HEX()` numeric conversion with MySQL while preserving string-byte semantics and legacy overload identities:

- treats BOOL as numeric 0/1;
- rounds DECIMAL64/128/256 exactly using one quotient/remainder decision;
- uses nearest-even conversion for ordinary REAL expressions; explicit REAL casts truncate in range and return BIGINT out-of-range errors beyond the signed integer domain, matching MySQL;
- rebinds prepared HEX arguments using execution-time source types without mutating the cached plan;
- preserves legacy overloads 0–7, including the original FLOAT executors 4/5;
- fences new HEX identities with **MORPCVersion74**; authoritative main reserves v69 for bounded RANK, v70 for row-dependent/unsigned CONV bases, v71 for checked integer arithmetic, v72 for corrected IP semantics, and v73 for widened DECIMAL SUM partial state;
- checks DEFAULT expressions before constant folding can erase new HEX identities;
- migrates legacy, non-folded expressions on execution-owned TableDef copies at the catalog-to-planner boundary once the negotiated protocol reaches v74;
- keeps a generation compiled under v73 internally consistent if the protocol reaches v74 before Run.

The MySQL REAL distinction is documented in https://github.com/matrixorigin/matrixone/issues/28457#issuecomment-5630537587.

### Folded DEFAULT migration — approved conservative scope

New HEX defaults retain their resolved expression in protobuf `Literal.Src`, preserving the resolved CAST type rather than reinterpreting creation SQL mode.

For legacy literals without provenance, migration accepts only a fully anchored, mode-independent lexical subset, then parses and binds it: `HEX(TRUE/FALSE)` and `HEX(CAST(signed decimal numeric literal AS DOUBLE/DECIMAL[(p[,s])]))`. It reconstructs the legacy executor result and requires it to match the stored value before replacing it with the corrected result. This restores DOUBLE `10 -> F`, BOOL `31 -> 1`, and exact DECIMAL conversion without unconditional reparsing of arbitrary defaults.

REAL aliases, quoted operands, nested arithmetic, other functions, and unmatched stored values remain unchanged. Old negative FLOAT-to-uint results are platform-dependent; the oracle uses the current platform's preserved executor and does not guess when it cannot match a historical value. These conservative exclusions are intentional; this is not a claim to migrate every possible folded expression.

All migration paths use copy-on-write: CHECK detaches its slice, wrapper and expression; generated/ON UPDATE/non-folded DEFAULT expressions detach their containing ColDef and wrappers; folded DEFAULT replacement detaches the column/default. Resolver shallow clones therefore never expose shared catalog expressions to mutation. Repeated migration is idempotent. REAL_AS_FLOAT defaults retain `1000000` with or without provenance.

### Service-level upgrade validation — completed at `165eb6cd05`

Built and ran the actual legacy binary from `be30a70fe5` (v68), created tables through SQL, stopped it gracefully, and started the PR binary (v69) against the same persistent data directory. No catalog records were synthesized or patched. Topology: one CN, one TN, one LOG; Go 1.26.4, macOS/arm64.

| Real service path | Verified outcome |
| --- | --- |
| Existing folded DOUBLE DEFAULT | New inserts change `10` to `F` |
| Existing positive DECIMAL DEFAULT | `20000000000000` to `20000000000001` |
| Existing negative DECIMAL DEFAULT | `0` to `FFDFFFFFFFFFFFFF` |
| Existing BOOL DEFAULT | `31` to `1` |
| Existing generated column and row-dependent DEFAULT | New rows for 14.5 produce `E` instead of `F` |
| Existing CHECK | Accepts corrected `E`; rejects legacy `F` with error 3819 |
| REAL_AS_FLOAT old/new defaults | Remain `1000000` across mode changes and restart |
| Second restart of the new binary | Inserts into the same old tables retain corrected defaults; existing rows remain unchanged |
| Live CN set to v68 using mo_ctl | CREATE TABLE with folded numeric HEX DEFAULT is rejected with the v69 requirement |
| Restore v69 | Same CREATE TABLE and INSERT succeed with `F` |

On a separate clean new instance, `func_hex.test` passed **twice in normal mo-tester comparison mode** (`-n -g -o -p`): **40/40, 100%, zero failed/ignored/abnormal** each time. Database teardown was checked after each run. Test-owned services, ports, data, and temporary legacy worktree/branch were cleaned up.

Full SQL transcripts, stage logs, reports, and the reproducible harness are retained in the author's worktree. These are single-node service upgrade results; they are not a claim of a live mixed-version multi-CN rollout test.

### Column metadata preservation follow-up

HEX COW now copies the complete ColDef value and complete Default/OnUpdate/GeneratedCol wrappers, deep-copying only expression trees. It no longer uses the incomplete global DeepCopyColDef helper. Regression tests compare the entire execution column against an independent protobuf-derived expectation, allowing only the target overload to change, for all three wrappers with both legacy and current HEX identities. NotNull, LowCard, Headers, Header, other metadata and catalog immutability are preserved.

The six cases failed before the fix and pass afterward. Focused race/vet and incremental lint passed. After merging main `7ffce04a09`, full function/plan/compile/frontend package tests also passed. The service-level evidence above records its exact previously validated revision; this metadata-specific follow-up uses the new full-column and ownership oracles.

### Latest main/CI follow-up — main `4cd720e0ce`

The latest Ubuntu/x86 UT run again had no test/race/panic failure and was externally terminated at the approximately 70-minute boundary while embedded-cluster work remained; completed packages were pass. This is the tracked #27909 runner/suite-budget condition, not a HEX assertion failure. Main was merged per CI policy. Upstream now owns v73 for widened DECIMAL SUM partial state, so HEX moved to v74 across planner/catalog migration, cached execution, remote sender/receiver, and all boundary tests; v73 rejects identities 8–14 and v74 accepts them.

After conflict resolution, full defines/container-types/function/plan/compile/frontend packages passed. Focused HEX race/vet, incremental lint (0 issues), and Go 1.26.4 formatter/diff checks passed. No unrelated CI test or product behavior was modified.

### Explicit REAL exact lower boundary follow-up

MySQL rejects explicit REAL conversion at `LLONG_MIN` inclusively. The explicit FLOAT32/FLOAT64 HEX overloads now use `converted <= float64(math.MinInt64)` while ordinary REAL keeps its established saturation behavior. Tests reject exact `-2^63` through both executors (it is exactly representable in each), retain positive/negative far-overflow coverage, and prove the next representable float64 toward zero still succeeds with two's-complement HEX. Masked overflow remains non-evaluated.

Focused normal and race/vet tests passed, the full function package passed, incremental lint reported 0 issues, and Go 1.26.4 formatting/diff checks passed. Main `b347a62a30` was already included. No service/BVT rerun was needed for this executor-boundary correction.

### Exact-head ownership race stress — `fd86276825`

The review-named test was selected exactly and measured with Go 1.26.4 using the repository CGo wrapper:

```text
GOTOOLCHAIN=go1.26.4 mo-cgo-test -mod=readonly -race -json -count=1 -timeout=120s -run '^TestHexCatalogMigrationOwnsAllExpressionWrappers$' ./pkg/sql/plan
```

Its terminal JSON events were `run` then `pass`, with test elapsed **T=0.01s** and exit 0. Under the adaptive race protocol's default **B=30s**, `floor(B/T)=3000`, clamped to **N=100**. The exact-test single-process stress then passed:

```text
GOTOOLCHAIN=go1.26.4 mo-cgo-test -mod=readonly -race -count=100 -timeout=120s -run '^TestHexCatalogMigrationOwnsAllExpressionWrappers$' ./pkg/sql/plan
ok .../pkg/sql/plan 4.698s; exit 0; wall 22.88s
```

Because this covers production catalog ownership/shared state, the owning package was also run once under race and passed: `./pkg/sql/plan` reported 40.368s, exit 0, wall 47.97s. All evidence is tied to exact PR head `fd86276825af12c3c0c395f6191e090ceddd8c53`; no code changed for this validation-only blocker.

### Latest CI/base synchronization — main `b347a62a30`

Ubuntu/x86 race UT attempts 1 and 2 had no Go test failure events. Both were externally cancelled at the same approximately 70-minute job boundary while resource-heavy/plan work remained; completed engine shards and all reported packages passed. This matches the outer-suite timeout tracked by #27909, so no unrelated test/product behavior was changed. The latest main was merged per failed-CI policy; HEX remains v73 with no new protocol allocation conflict. Full defines/function/plan/compile/frontend package tests passed locally with Go 1.26.4 before pushing the new HEAD.

### Latest review and merge — main `2ea4ec2e6e`

Explicit FLOAT32/FLOAT64 CAST overloads now return BIGINT out-of-range errors for positive and negative values outside the signed integer domain (`±1e20` coverage); ordinary REAL and DECIMAL saturation contracts remain unchanged. A masked overflow row is also proven not to fail short-circuit execution. Latest main allocated v70–v72, so HEX moved to v73 across planner/catalog, cached execution, remote sender/receiver and boundary tests; v72 rejects identities 8–14 and v73 accepts them.

After the merge, full defines/function/plan/compile/frontend packages passed with Go 1.26.4. Focused HEX race/vet and incremental lint passed. The first full function package run exposed only an incorrect selection mask in the new test; after correcting the test oracle, the focused and full rerun passed. No complete service upgrade was rerun for this follow-up; service evidence below is explicitly tied to its historical revision and protocol allocation.

### Unassignable legacy DEFAULT follow-up — `2001e18a42`

If the corrected HEX result cannot be assigned to the original column type, catalog migration now retains the legacy DEFAULT rather than returning an error from the generic resolver. For example, DECIMAL(5,1) keeps the stored 10.0 when corrected HEX would produce the unassignable string `F`. This deliberately preserves the old default for subsequent inserts too; it does not defer an error to reads or partially publish a replacement.

DECIMAL64 and DECIMAL128 tests first reproduced the failure using values generated through the preserved legacy executor and assignment folding, then passed after the fix. A real frontend ResolveById regression also passes and checks repeated resolution and catalog immutability. After merging main `f9a4444774`, full plan/compile/frontend package tests, focused race/vet, and incremental lint (0 issues) passed at `abeaef205b`. This follow-up did not rerun the full service upgrade; earlier service evidence retains its original revision below/above.

### Latest protocol merge validation — main `a72a224ce8`

Resolved the v69 allocation conflict by preserving upstream RANK v69 and moving HEX gates to v70. Reviewed the remote encode/decode merge: both features retain their independent fences. Full defines/function/plan/compile/frontend package tests passed with Go 1.26.4; focused HEX tests passed with race and vet=all, incremental lint reported 0 issues, and formatter/diff checks passed. HEX overloads 8–14 are rejected at v69 and accepted at v70, including planner pre-fold admission, cached execution, and sender/receiver checks. Older-version controls remain.

The service upgrade results below/above are historical evidence at `165eb6cd05`, not a rerun against this merge or proof of the newly numbered service gate.

### Validation evidence

After merging main `be30a70fe5`: full `pkg/defines`, `pkg/container/types`, `pkg/sql/plan/function`, `pkg/sql/plan`, `pkg/sql/compile`, and `pkg/frontend` tests passed with Go 1.26.4. Focused migration/protocol tests passed with race and vet=all; incremental lint reported 0 issues; make build passed. Tests cover concurrent resolver shallow clones, immutability of CHECK/generated/ON UPDATE/non-folded DEFAULT wrappers, protobuf round-trip, repeated migration, SQL-mode controls, and execution of a migrated folded DEFAULT. Remote tests reject all new HEX identities 8–14 under v68 and accept them under v69; old identities remain supported.

The initial parallel race/lint attempt exhausted temporary disk space; serial reruns passed after temporary build files were released. The previously reported `331/404` coverage is historical. Service upgrade and two normal mo-tester runs at `165eb6cd05` are recorded separately above.

### CODE REVIEW

**Review by aunjgr — APPROVED** [review]
Reviewed revision: `e8ac7cf363af212404f4a66b1cee6ee0ee8403e9`.

Independent review: no blocking finding. Checked exact DECIMAL rounding and signed boundaries, ordinary versus explicit REAL cast semantics, BOOL/string/prepared source dispatch, immutable catalog migration and folded-default fallback, and Compile/Run/remote admission.

Legacy overload identities remain stable; corrected numeric identities use a fresh v70 allocation relative to reviewed authoritative main v69. The migration preserves unassignable legacy defaults rather than making table resolution fail.

Reused exact-head CI run 34842091321; no local kernel/BVT rerun.

**Review by XuPeng-SH — CHANGES_REQUESTED** [review]
GPT-6 medium deep review at exact head e8ac7cf363af212404f4a66b1cee6ee0ee8403e9.

**P2 — explicit REAL conversion still diverges from MySQL.** `HEX(CAST(1e20 AS DOUBLE))` reaches signed integer conversion in `pkg/sql/plan/function/func_unary.go:7643–7646`; the current path saturates rather than returning the MySQL-compatible overflow error. Please add explicit REAL boundary cases and preserve the expected error behavior.

The exact-head CI is green. Available artifacts do not expose individual timings for the new tests, so the 45-second per-test limit is not certified.

**Review by XuPeng-SH — CHANGES_REQUESTED** [review]
GPT-6 medium deep review at exact head fd86276825af12c3c0c395f6191e090ceddd8c53.

P2 — focused race-stress evidence is not established for the catalog migration ownership path at pkg/sql/plan/hex_catalog_ownership_test.go:138. Concurrent resolvers shallow-clone a catalog containing aliased CHECK, DEFAULT, generated-column, and ON UPDATE expressions; the migration must detach every writable object. Exact-head CI proves a race-suite run, but does not provide the measured, individually selected repetition of TestHexCatalogMigrationOwnsAllExpressionWrappers.

Please provide the named test's terminal race result, measured duration, derived repetition count, and repeated-run result tied to this head or a proven equivalent revision. This is a validation-evidence blocker, not a demonstrated race.

File: pkg/sql/plan/function/func_unary.go [hunk 1]
```diff
@@ -7720,20 +7720,173 @@ func HexUint64(ivecs []*vector.Vector, result vector.FunctionResultWrapper, proc
...
+
+func HexExplicitFloat64(ivecs []*vector.Vector, result vector.FunctionResultWrapper, proc *process.Process, length int, selectList *FunctionSelectList) error {
+	return hexFloat[float64](ivecs, result, proc, length, selectList, true)
+}
+
+func hexFloat[T constraints.Float](ivecs []*vector.Vector, result vector.FunctionResultWrapper, proc *process.Process, length int, selectList *FunctionSelectList, truncate bool) error {
+	if truncate {
+		return opUnaryFixedToStrWithErrorCheck[T](ivecs, result, proc, length, func(value T) (string, error) {
+			converted := math.Trunc(float64(value))
+			if converted >= float64(math.MaxInt64) || converted < float64(math.MinInt64) {
```
Comment (aptend) [inline pkg/sql/plan/function/func_unary.go]: P2: MySQL's signed conversion rejects `res <= LLONG_MIN`, but this lower predicate only rejects values strictly below it. Consequently `SELECT HEX(CAST(-9223372036854775808 AS DOUBLE));` returns `8000000000000000` here while MySQL raises integer overflow. I reproduced this directly on the exact head with `float64(math.MinInt64)`. Please make the lower check inclusive and add exact-boundary coverage (including the FLOAT32 executor, since `-2^63` is exactly representable there).

**Review by aptend — CHANGES_REQUESTED** [review]
Deep re-review of exact head `fd86276825af12c3c0c395f6191e090ceddd8c53`. I read the complete review/comment history (there are no inline threads), compared the increment since the prior reviewed head `e8ac7cf363af212404f4a66b1cee6ee0ee8403e9`, and checked the complete diff. The prior `1e20` explicit-REAL saturation issue is fixed, but the exact signed lower boundary still diverges from MySQL, so this revision remains blocked.

Reproduction on this head: a review-only unit case passed `float64(math.MinInt64)` to `HexExplicitFloat64` and expected the MySQL overflow error; it failed with `expected to run failed, but run succeed with no error`. The temporary test was removed afterward.

Validation: `make thirdparties`; the focused boundary test above; `go test -race -count=1 -run '^TestHexCatalogMigrationOwnsAllExpressionWrappers$' ./pkg/sql/plan`; and the same named race test with `-count=10`. Both race runs passed, independently clearing the catalog-ownership evidence concern. Exact-head required CI is green.

**Review by XuPeng-SH — APPROVED** [review]
GPT-6 medium deep re-review APPROVE at exact head 68cad45fe9af4657cabf7669e12eaf3b3ec08055\n\nAll changed hunks and applicable correctness, ownership, lifecycle, liveness, boundedness, compatibility, and test closures reviewed. Prior findings are resolved: explicit REAL overflow includes the exact lower boundary, and the catalog ownership race evidence remains applicable. Exact-head build, UT/race, static checks, and BVT passed; func_hex.test passed 40/40 with zero failures or ignored cases.
